tidepool_data_science_metrics/cgm/cgm.py

```diff
@@ -0,0 +1,187 @@
+
+
+def percent_values_by_range(
+    bg_values, lower_threshold: int, upper_threshold: int, round_val=2
+):
+    """
+    Calculate the percent of values that match has a bg within the lower and upper threshold.
+    The lower and upper values will be included in the range to calculate on.
+
+    Parameters
+    ----------
+    bg_values : ndarray
+        1D array containing data with `int` type.
+    lower_threshold : int
+        The the lower value in the range to calculate on.
+    upper_threshold : int
+        zThe the upper value in the range to calculate on.
+    round_val : int
+        The number of digits to round the result to.
+
+    Returns
+    -------
+    int
+        The percent value by range.
+    """
+
+    calc_low_thresh, calc_upper_thresh = _validate_input(
+        lower_threshold, upper_threshold
+    )
+    results = round(
+        np.where(
+            (bg_values <= calc_upper_thresh) & (bg_values >= calc_low_thresh), 1, 0
+        ).sum()
+        / bg_values.size
+        * 100,
+        round_val,
+    )
+    return results
+
+
+def percent_time_in_range(
+    bg_values, lower_threshold: int, upper_threshold: int, round_val=2, time_delta=5
+):
+
+    calc_low_thresh, calc_upper_thresh = _validate_input(
+        lower_threshold, upper_threshold
+    )
+    bg_df = bg_values["bg_values"]
+    bg_np = bg_df.to_numpy()
+    in_range = np.count_nonzero((bg_np > calc_low_thresh) & (bg_np < calc_upper_thresh))
+    val_count = bg_df.count()
+    return round(in_range / val_count * 100, round_val)
+
+
+def episodes(
+    bg_values_df, episodes_threshold: int, min_ct_per_ep=3, min_duration=5, round_val=2
+):
+    """
+    Calculate the number of episodes for a given set of glucose values based on provided thresholds.
+
+    Parameters
+    ----------
+    bg_values : ndarray
+        1D array containing data with `int` type.
+    lower_threshold : int
+        The the lower value in the range to calculate on.
+    upper_threshold : int
+        zThe the upper value in the range to calculate on.
+    round_val : int
+        The number of digits to round the result to.
+
+    Returns
+    -------
+    int
+        The number of episodes matching input specifications.
+    """
+    bg_values_df.loc[(bg_values_df["values"] < episodes_threshold), "episode"] = 1
+    bg_values_df.loc[(bg_values_df["values"] >= episodes_threshold), "episode"] = 0
+    bg_values_df["group"] = 0
+    bg_values_df["group"][
+        (
+            (bg_values_df.episode == 1)
+            & (bg_values_df.episode.shift(1) == 0)
+            & (bg_values_df.episode.shift(-1) == 1)
+            & (bg_values_df.episode.shift(-(min_ct_per_ep - 1)) == 1)
+        )
+    ] = 1
+    bg_values_df["group"][
+        (
+            (bg_values_df.episode == 1)
+            & (bg_values_df.index == 0)
+            & (bg_values_df.episode.shift(-1) == 1)
+            & (bg_values_df.episode.shift(-2) == 1)
+        )
+    ] = 1
```

Comment:
I have a general idea of what's going on here, but it'd be great to add plain language comments detailing each of these masks and calls. Updating the docstring would probably help too, but I know this is a work in progress